test/utils/ruleset-test-harness.ts

```diff
@@ -0,0 +1,151 @@
+  setUp: (rules: any, dataLayer: any) => {
+    setupGlobals(
+      Object.keys(dataLayer).map((key) => [key, dataLayer[key]]),
+    );
+    ExpectObserver.getInstance().create({ rules });
+    return Promise.resolve();
+  },
+
+  tearDown: () => {
+    ExpectObserver.getInstance().cleanup();
+    return Promise.resolve();
+  },
+
+  execute: (action: (args: any[]) => void, args?: any[]) => {
+    action(args || []);
+    return Promise.resolve();
+  },
+
+  popEvent: async (timeoutMs: number = 1000) => new Promise<any[]>((resolve, reject) => {
+    const startTime = new Date().getTime();
+
+    const checkForEvents = setInterval(() => {
+      const currentTime = new Date().getTime();
+      if (currentTime - startTime >= timeoutMs) {
+        clearInterval(checkForEvents);
+        reject(new Error(`Timeout ${timeoutMs} exceeded waiting for FS.event call.`));
+      }
+
+      if (expectGlobal('FS').callQueues.event.length) {
+        clearInterval(checkForEvents);
+        resolve(expectFS('event'));
+      }
+    }, 50);
+  }),
+};
+
+class BrowserTestHarness implements RulesetTestHarness {
+  private browser: Browser = null!;
+
+  private page: Page = null!;
+
+  constructor(private readonly browserPromise: Promise<Browser>) {
+    if (!dloScriptSrc) {
+      throw new Error(
+        'PLAYWRIGHT_DLO_SCRIPT_SRC must be configured with a DLO library script location for browser tests.',
+      );
+    }
+  }
+
+  async setUp(rules: any, dataLayer: any) {
+    this.browser = await this.browserPromise;
+    this.page = await this.browser.newPage();
+
+    await this.page.evaluate(([localRules, localDataLayer, localDloScriptSrc]) => {
+      // This allows node and playwright tests to reference the same global/window name.
+      const globalThis: any = window;
+
+      globalThis._dlo_rules = localRules;
+
+      Object.keys(localDataLayer).forEach((key) => {
+        globalThis[key] = localDataLayer[key];
+      });
+
+      globalThis.events = [];
+
+      globalThis.FS = {
+        event: (...args: any) => {
+          globalThis.events.push(args);
+        },
+      };
+
+      const dloScriptTag = document.createElement('script');
+      dloScriptTag.src = localDloScriptSrc;
+      document.body.appendChild(dloScriptTag);
+    }, [rules, dataLayer, dloScriptSrc]);
+
+    // Wait for DLO to initialize.
+    await this.page.waitForFunction(() => (window as any)._dlo_observer, undefined, { timeout: 1000 });
+  }
+
+  async tearDown() {
+    await this.page.close();
+  }
+
+  async execute(action: (args: any[]) => void, args?: any[]) {
+    await this.page.evaluate(action, args || []);
+  }
+
+  async popEvent(timeoutMs: number = 1000) {
+    await this.page.waitForFunction(() => (window as any).events.length, undefined, { timeout: timeoutMs });
+    return this.page.evaluate(() => (window as any).events.pop());
+  }
+}
+
+export const getRulesetTestEnvironments = (): RulesetTestEnvironment[] => {
```

Comment:
TLDR: this looks like a factory method that could use some parameters.

if my understanding is correct this is the initializer for the tests. if so we could give this method a parameter of options that will help configure these settings. 

use case: some tests for us on eco-playwright don't work on certain browsers, we may need to disable a browser/environment from running tests. i can't _currently_ imagine that would happen for DLO.
use case: as a developer, i don't want to iterate on all tests, when developing tests it might help if i can quickly specify what browsers to run, global timeout logic or settings.

not sure where this falls on overengineering vs futureproofing, wdyt?

Reply:
I think you've identified a good entry point for future config options -- whether defined here explicitly as args or via other env vars passed in on `process.env` (`DLO_RUN_BROWSER_TESTS` influences the return value here already). I'm not sure what configuration values would be useful today given we'd expect to primarily run these tests for release candidates in staging and that frequency is low. How would you feel about waiting to see what config options make sense in the future?

Reply:
i'm okay with waiting 👍